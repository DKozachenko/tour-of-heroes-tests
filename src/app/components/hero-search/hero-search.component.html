<div
  id="search-component"
  automation-id="search-component"
  pw-automation-id="search-component"
>
  <label
    for="search-box"
    automation-id="search-label"
    pw-automation-id="search-label"
    >Hero Search</label
  >
  <input
    #searchBox
    id="search-box"
    (input)="search(searchBox.value)"
    automation-id="search-input"
    pw-automation-id="search-input"
  />

  <ul
    class="search-result"
    automation-id="search-result"
    pw-automation-id="search-result"
  >
    <li *ngFor="let hero of heroes$ | async">
      <a
        routerLink="/detail/{{ hero.id }}"
        automation-id="search-link"
        pw-automation-id="search-link"
      >
        {{ hero.name }}
      </a>
    </li>
  </ul>
</div>
