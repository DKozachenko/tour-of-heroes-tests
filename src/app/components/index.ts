export { DashboardComponent } from './dashboard/dashboard.component';
export { HeroDetailComponent } from './hero-detail/hero-detail.component';
export { HeroSearchComponent } from './hero-search/hero-search.component';
export { HeroesComponent } from './heroes/heroes.component';
export { MessagesComponent } from './messages/messages.component';
export { AppComponent } from './app/app.component';
