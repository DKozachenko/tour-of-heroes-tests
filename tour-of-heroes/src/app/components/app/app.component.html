<h1 automation-id="heading">Tour of Heroes</h1>
<nav automation-id="navigation">
  <a routerLink="/dashboard" automation-id="navigation-link">Dashboard</a>
  <a routerLink="/heroes" automation-id="navigation-link">Heroes</a>
</nav>
<router-outlet></router-outlet>
<app-messages automation-id="messages"></app-messages>
