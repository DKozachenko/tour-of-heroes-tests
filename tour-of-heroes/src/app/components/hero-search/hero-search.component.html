<div id="search-component" automation-id="search-component">
  <label for="search-box" automation-id="search-label">Hero Search</label>
  <input
    #searchBox
    id="search-box"
    (input)="search(searchBox.value)"
    automation-id="search-input"
    pw-automation-id="search-input"
  />

  <ul class="search-result" automation-id="search-result">
    <li *ngFor="let hero of heroes$ | async">
      <a
        routerLink="/detail/{{ hero.id }}"
        automation-id="search-link"
        pw-automation-id="search-link"
      >
        {{ hero.name }}
      </a>
    </li>
  </ul>
</div>
