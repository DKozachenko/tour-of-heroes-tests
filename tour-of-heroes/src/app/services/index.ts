export { HeroService } from './hero/hero.service';
export {
  InMemoryDataService,
  HEROES,
} from './in-memory-data/in-memory-data.service';
export { MessageService } from './message/message.service';
