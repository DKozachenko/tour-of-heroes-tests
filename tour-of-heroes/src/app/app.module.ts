import { BrowserModule } from '@angular/platform-browser';
import { FormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { AppComponent } from './app.component';
import { AppRoutingModule } from './app-routing.module';
import { HttpClientInMemoryWebApiModule } from 'angular-in-memory-web-api';
import { InMemoryDataService } from './services';

import {
  DashboardComponent,
  HeroDetailComponent,
  HeroesComponent,
  HeroSearchComponent,
  MessagesComponent,
} from './components';

@NgModule({
  imports: [
    BrowserModule,
    FormsModule,
    AppRoutingModule,
    HttpClientModule,

    // The HttpClientInMemoryWebApiModule module intercepts HTTP requests
    // and returns simulated server responses.
    // Remove it when a real server is ready to receive requests.
    HttpClientInMemoryWebApiModule.forRoot(InMemoryDataService, {
      dataEncapsulation: false,
    }),
  ],
  declarations: [
    AppComponent,
    DashboardComponent,
    HeroesComponent,
    HeroDetailComponent,
    MessagesComponent,
    HeroSearchComponent,
  ],
  bootstrap: [AppComponent],
})
export class AppModule {}
