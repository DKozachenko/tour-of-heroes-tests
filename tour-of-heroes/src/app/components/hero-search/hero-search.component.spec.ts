import { Router, RouterLink, RouterModule } from '@angular/router';
import { DebugElement } from '@angular/core';
import { By } from '@angular/platform-browser';
import { Location } from '@angular/common';
import { RouterTestingModule } from '@angular/router/testing';
import { fakeAsync, discardPeriodicTasks, tick } from '@angular/core/testing';
import {
  MockBuilder,
  MockRender,
  MockedComponentFixture,
  NG_MOCKS_ROOT_PROVIDERS,
} from 'ng-mocks';
import { of } from 'rxjs';
import { mock, instance, when, anyString } from 'ts-mockito';
import { cold } from 'jest-marbles';
import { HeroService, HEROES } from '../../services';
import { Hero } from '../../models';
import { HeroSearchComponent } from './hero-search.component';
import { AppModule } from '../../app.module';

class PageObject {
  private fixtureDebugElement: DebugElement;

  constructor(fixture: MockedComponentFixture<HeroSearchComponent>) {
    this.fixtureDebugElement = fixture.debugElement;
  }

  get divWrapper(): DebugElement {
    return this.fixtureDebugElement.query(By.css('#search-component'));
  }

  get label(): DebugElement {
    return this.fixtureDebugElement.query(By.css('label'));
  }

  get input(): DebugElement {
    return this.fixtureDebugElement.query(By.css('input'));
  }

  get searchResult(): DebugElement {
    return this.fixtureDebugElement.query(By.css('.search-result'));
  }

  get heroLinks(): DebugElement[] {
    return this.searchResult.queryAll(By.css('li'));
  }
}

describe('HeroSearchComponent', () => {
  let mockHeroService: HeroService;

  beforeEach(() => {
    mockHeroService = mock(HeroService);

    return MockBuilder(HeroSearchComponent, AppModule).mock(
      HeroService,
      instance(mockHeroService)
    );
  });

  function createFixture(): MockedComponentFixture<HeroSearchComponent> {
    return MockRender(HeroSearchComponent);
  }

  function createComponent(): HeroSearchComponent {
    return createFixture().point.componentInstance;
  }

  function mockCalls(): void {
    when(mockHeroService.searchHeroes(anyString())).thenReturn(of([]));
  }

  describe('Functional', () => {
    it('should define "heroes$" property as empty observable while creating', () => {
      mockCalls();
      const component = createComponent();
      expect(component.heroes$).toBeDefined();
      expect(component.heroes$).toBeObservable(cold(''));
    });

    it('should not change "heroes$" property values when "search" method called if less than 300ms has passed', fakeAsync(() => {
      const mockSearchedHeroes = [HEROES[0], HEROES[1]];
      when(mockHeroService.searchHeroes(anyString())).thenReturn(
        of(mockSearchedHeroes)
      );
      let heroValues: Hero[] = [];
      const component = createComponent();
      component.heroes$.subscribe((newHeroes) => (heroValues = newHeroes));
      component.search(anyString());

      tick();
      expect(heroValues).toStrictEqual([]);
      // Should use for clearing queue
      // https://stackoverflow.com/questions/43060886/angular-2-fakeasync-waiting-for-timeout-in-a-function-using-tick
      discardPeriodicTasks();
    }));

    it('should not change "heroes$" property values when "search" method called if new value is the same as previous (more than 300ms has passed)', fakeAsync(() => {
      const mockOldSearchedHeroes = [HEROES[0], HEROES[1]];
      when(mockHeroService.searchHeroes(anyString())).thenReturn(
        of(mockOldSearchedHeroes)
      );

      let heroValues: Hero[] = [];
      const component = createComponent();
      component.heroes$.subscribe((newHeroes) => (heroValues = newHeroes));

      const oldSearchValue1 = 'test 1';
      const newSearchValue2 = 'test 2';
      component.search(oldSearchValue1);
      tick(500);
      expect(heroValues).toStrictEqual(mockOldSearchedHeroes);

      const mockNewSearchedHeroes = [HEROES[1], HEROES[2]];
      when(mockHeroService.searchHeroes(anyString())).thenReturn(
        of(mockNewSearchedHeroes)
      );
      component.search(newSearchValue2);
      component.search(oldSearchValue1);
      tick(500);
      expect(heroValues).toStrictEqual(mockOldSearchedHeroes);
    }));

    it('should change "heroes$" property values when "search" method called (more than 300ms has passed and new value emitted)', fakeAsync(() => {
      const mockOldSearchedHeroes = [HEROES[0], HEROES[1]];

      when(mockHeroService.searchHeroes(anyString())).thenReturn(
        of(mockOldSearchedHeroes)
      );

      let heroValues: Hero[] = [];
      const component = createComponent();
      component.heroes$.subscribe((newHeroes) => (heroValues = newHeroes));

      const oldSearchValue1 = 'test 1';
      const newSearchValue2 = 'test 2';
      component.search(oldSearchValue1);
      tick(500);
      expect(heroValues).toStrictEqual(mockOldSearchedHeroes);

      const mockNewSearchedHeroes = [HEROES[1], HEROES[2]];
      when(mockHeroService.searchHeroes(anyString())).thenReturn(
        of(mockNewSearchedHeroes)
      );
      component.search(newSearchValue2);
      component.search(oldSearchValue1);
      component.search(newSearchValue2);
      tick(500);
      expect(heroValues).toStrictEqual(mockNewSearchedHeroes);
    }));
  });

  describe('Layout', () => {
    it('should contain div wrapper', () => {
      mockCalls();
      const fixture = createFixture();
      const pageObject = new PageObject(fixture);
      expect(pageObject.divWrapper).not.toBeNull();
      expect(pageObject.divWrapper.nativeElement).toMatchSnapshot();
    });

    it('should contain label and input for typing', () => {
      mockCalls();
      const fixture = createFixture();
      const pageObject = new PageObject(fixture);
      expect(pageObject.label).not.toBeNull();
      expect(pageObject.label.nativeElement).toMatchSnapshot();
      expect(pageObject.input).not.toBeNull();
      expect(pageObject.input.nativeElement).toMatchSnapshot();
    });

    it('should call "search" component method while typing with input value', () => {
      mockCalls();
      const fixture = createFixture();
      const component = fixture.point.componentInstance;

      const pageObject = new PageObject(fixture);
      const spyOnSearch = jest.spyOn(component, 'search');

      const newInputValue = 'new input value';
      pageObject.input.nativeElement.value = newInputValue;
      pageObject.input.triggerEventHandler('input');
      expect(spyOnSearch).toHaveBeenCalled();
      expect(spyOnSearch).toHaveBeenCalledWith(newInputValue);
    });

    describe('hero list', () => {
      it('should contain hero list with hero names and hero id as routerLink attribute', () => {
        mockCalls();
        const fixture = createFixture();
        const component = fixture.point.componentInstance;
        const mockHeroes = HEROES.slice(0, 3);
        component.heroes$ = of(mockHeroes);
        fixture.detectChanges();

        const pageObject = new PageObject(fixture);
        expect(pageObject.searchResult).not.toBeNull();
        expect(pageObject.searchResult.nativeElement).toMatchSnapshot();
        expect(pageObject.heroLinks).toHaveLength(mockHeroes.length);

        for (let i = 0; i < pageObject.heroLinks.length; ++i) {
          const heroLink = pageObject.heroLinks[i];
          const hero = mockHeroes[i];
          expect(heroLink.nativeElement.textContent).toContain(hero.name);
          expect(heroLink.nativeElement).toMatchSnapshot();
        }
      });

      it('should change hero list if new value has typed in input (more than 300ms has passed and new value emitted)', fakeAsync(() => {
        const mockOldSearchedHeroes = [HEROES[0], HEROES[1]];

        when(mockHeroService.searchHeroes(anyString())).thenReturn(
          of(mockOldSearchedHeroes)
        );

        const fixture = createFixture();
        const component = fixture.point.componentInstance;

        const oldSearchValue1 = 'test 1';
        component.search(oldSearchValue1);
        tick(500);
        fixture.detectChanges();

        const pageObject = new PageObject(fixture);
        expect(pageObject.searchResult).not.toBeNull();
        expect(pageObject.heroLinks).toHaveLength(mockOldSearchedHeroes.length);

        for (let i = 0; i < pageObject.heroLinks.length; ++i) {
          const heroLink = pageObject.heroLinks[i];
          const hero = mockOldSearchedHeroes[i];
          expect(heroLink.nativeElement.textContent).toContain(hero.name);
        }

        const mockNewSearchedHeroes = [HEROES[2], HEROES[3], HEROES[4]];
        when(mockHeroService.searchHeroes(anyString())).thenReturn(
          of(mockNewSearchedHeroes)
        );

        const newSearchValue2 = 'test 2';
        component.search(newSearchValue2);
        tick(500);
        fixture.detectChanges();

        expect(pageObject.searchResult).not.toBeNull();
        expect(pageObject.heroLinks).toHaveLength(mockNewSearchedHeroes.length);

        for (let i = 0; i < pageObject.heroLinks.length; ++i) {
          const heroLink = pageObject.heroLinks[i];
          const hero = mockNewSearchedHeroes[i];
          expect(heroLink.nativeElement.textContent).toContain(hero.name);
        }
      }));
    });
  });
});

describe('HeroSearchComponent:Routing', () => {
  let mockHeroService: HeroService;

  beforeEach(() => {
    mockHeroService = mock(HeroService);

    return MockBuilder(
      [
        HeroSearchComponent,
        RouterModule,
        RouterTestingModule.withRoutes([]),
        NG_MOCKS_ROOT_PROVIDERS,
      ],
      AppModule
    ).mock(HeroService, instance(mockHeroService));
  });

  function createFixture(): MockedComponentFixture<HeroSearchComponent> {
    return MockRender(HeroSearchComponent);
  }

  function mockCalls(): void {
    when(mockHeroService.searchHeroes(anyString())).thenReturn(of([]));
  }

  it('should change location if appropriate hero option has clicked', fakeAsync(() => {
    mockCalls();
    const fixture = createFixture();
    const router = fixture.point.injector.get(Router);
    const location = fixture.point.injector.get(Location);

    if (fixture.ngZone) {
      fixture.ngZone.run(() => router.initialNavigation());
      tick();
    }

    const pageObject = new PageObject(fixture);
    for (let i = 0; i < pageObject.heroLinks.length; ++i) {
      const heroLink = pageObject.heroLinks[i];
      if (fixture.ngZone) {
        fixture.ngZone.run(() => {
          heroLink.triggerEventHandler('click', {
            button: 0,
          });
        });
        tick();
      }

      expect(location.path()).toBe(heroLink.injector.get(RouterLink).href);
    }
  }));
});
