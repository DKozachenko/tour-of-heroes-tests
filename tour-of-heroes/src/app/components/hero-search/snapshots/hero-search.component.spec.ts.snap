// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroSearchComponent Layout hero list should contain hero list with hero names and hero id as routerLink attribute 1`] = `
<ul
  automation-id="search-result"
  class="search-result"
>
  <li>
    <a
      automation-id="search-link"
    >
       Dr. Nice 
    </a>
    
  </li>
  <li>
    <a
      automation-id="search-link"
    >
       Bombasto 
    </a>
    
  </li>
  <li>
    <a
      automation-id="search-link"
    >
       Celeritas 
    </a>
    
  </li>
  
</ul>
`;

exports[`HeroSearchComponent Layout hero list should contain hero list with hero names and hero id as routerLink attribute 2`] = `
<a
  automation-id="search-link"
>
   Dr. Nice 
</a>
`;

exports[`HeroSearchComponent Layout hero list should contain hero list with hero names and hero id as routerLink attribute 3`] = `
<a
  automation-id="search-link"
>
   Bombasto 
</a>
`;

exports[`HeroSearchComponent Layout hero list should contain hero list with hero names and hero id as routerLink attribute 4`] = `
<a
  automation-id="search-link"
>
   Celeritas 
</a>
`;

exports[`HeroSearchComponent Layout should contain div wrapper 1`] = `
<div
  automation-id="search-component"
  id="search-component"
>
  <label
    automation-id="search-label"
    for="search-box"
  >
    Hero Search
  </label>
  <input
    automation-id="search-input"
    id="search-box"
  />
  <ul
    automation-id="search-result"
    class="search-result"
  >
    
  </ul>
</div>
`;

exports[`HeroSearchComponent Layout should contain label and input for typing 1`] = `
<label
  automation-id="search-label"
  for="search-box"
>
  Hero Search
</label>
`;

exports[`HeroSearchComponent Layout should contain label and input for typing 2`] = `
<input
  automation-id="search-input"
  id="search-box"
/>
`;
